lib/core/class-groups-user-group.php

```diff
@@ -139,6 +139,19 @@ public static function read( $user_id, $group_id ) {
 		global $wpdb;
 		$result = false;
 
+		/**
+		 * Allow short circuit the results of a user-group relation
+		 * @param  bool $result  (false)
+		 * @param  int  $user_id
+		 * @param  int  $group_id
+		 * @return bool|int  false or group object
+		 */
+		if ( $result = apply_filters( 'groups_user_is_member', $result, $user_id, $group_id ) ) {
```

Comment:
reversed approach - should be applied on result instead (before returned)

Reply:
This method is essentially a shortcut which will prevent any unnecessary logic to run and take up resources.
The same method is used in WP core many time.

Reply:
This filter has been added with a slightly different signature https://github.com/itthinx/groups/blob/56957ae3b296b83e0edc436e13641581c5d6130d/lib/core/class-groups-user.php#L388 and allows to filter the result. The 2.20.0 branch has also had various instances refactored that relied on the `Groups_User_Group::read(...)` method and which are now using the `Groups_User::user_is_member(...)` or `$group_user->is_member(...)` methods where appropriate so that the new filter kicks in. Extensions to Groups that were using the former method, should be updated to use the new ones.

As to the short-circuit filter suggested, which would allow to preemptively return a value before Groups polls its data, although this pattern is used in WordPress core, I see that it's basically done for two reasons: to evaluate something that is needed in a function for further processing (not overriding the actual processing of a return value), or to preemptively provide an alternative way to evaluate the return value (in the form of `pre_...` filters). For the latter, one could make a case for every single function to provide such a short-circuit filter for the sake of extensibility, but to me that's anti-pattern terrain. Instead, I would rather go the way of using OOP-based extensibility, for example, by introducing factories that allow to provide alternative classes to the default ones. Otherwise we would end up adding `pre_...` and `$result` filters on each and every method and property, adding overhead in processing for custom cases. I think that this could be handled better by using proper inheritance and factories.